### PR TITLE
File with include extractor

### LINKED ISSUE #489: [kythe] includes should emit file reference links
## Input test case

```systemverilog
package p;
`include "path/to/foo.svh"
endpackage
```

## Describe what is wrong or missing

The `"path/to/foo.svh"` should link to that named file.
I understand this is blocked on multi-file support.

### PR DESCRIPTION
**Pull request by MinaToma**
Fixes #489

### CODE REVIEW

File: verilog/tools/kythe/indexing_facts_tree_extractor_test.cc [hunk 1]
```diff
@@ -122,16 +131,15 @@ TEST(FactsTreeExtractor, ParseErrorTest) {
...
 
+    ScopedTestFile test_file(testing::TempDir(), code_text);
     const auto facts_tree =
-        ExtractOneFile(code_text, file_name, exit_status, parse_ok);
-    EXPECT_FALSE(parse_ok) << "code\n" << code_text;
-    EXPECT_EQ(exit_status, 1) << "code\n" << code_text;
+        ExtractFiles({std::string(test_file.filename())}, exit_status, "");
+    LOG(INFO) << facts_tree;
+    EXPECT_EQ(exit_status, 0) << "code\n" << code_text;
   }
```
Comment (MinaToma) [inline verilog/tools/kythe/indexing_facts_tree_extractor_test.cc:143]: @fangism how should we test this now? 
we don't have parse_ok and exist_status now.
I think this isn't a suitable response to the user now we should return something more meaningful.

File: verilog/tools/kythe/indexing_facts_tree_extractor.h [hunk 2]
```diff
@@ -87,6 +91,9 @@ class IndexingFactsTreeExtractor : public verible::TreeContextVisitor {
   // to macros and creates its corresponding facts tree.
   void ExtractMacroReference(const verible::SyntaxTreeLeaf& macro_identifier);
 
+  // Extracts Include statements and creates its corresponding fact tree.
+  void ExtractInclude(const verible::SyntaxTreeNode& preprocessor_include);
```
Comment (MinaToma) [inline verilog/tools/kythe/indexing_facts_tree_extractor.h:95]: No Tests added for this function yet.

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/my_module2.sv [hunk 3]
```diff
@@ -0,0 +1,28 @@
+`include "included-file-2.svh"
```
Comment (MinaToma) [inline verilog/tools/kythe/testdata/more_testdata/include_file_test/my_module2.sv:1]: @ivan444 is there some kind of verification for this line?
I tried to add `//- @` here but didn't know how to do it.
Do you know how to do this?

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 4]
```diff
@@ -37,34 +37,24 @@ ABSL_FLAG(bool, printkythefacts, false,
 ABSL_FLAG(std::string, output_path, "",
           "File path where to write the extracted Kythe facts in JSON format.");
 
-static int ExtractOneFile(
-    absl::string_view content, absl::string_view filename,
-    std::vector<verilog::kythe::IndexingFactNode>& indexing_facts_trees) {
+static int ExtractFiles(std::vector<std::string> ordered_file_list,
```
Comment (fangism) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:40]: `const& ordered_file_list`

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 5]
```diff
@@ -73,42 +63,43 @@ static void ExtractKytheFacts(
...
-    int file_status = ExtractOneFile(content, filename, indexing_facts_trees);
-    exit_status = std::max(exit_status, file_status);
-  }
+  std::vector<std::string> files_names;
+  std::string filename;
 
-  ExtractKytheFacts(indexing_facts_trees);
+  std::stringstream stream(content);
+  while (stream >> filename) {
+    files_names.push_back(filename);
```
Comment (fangism) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:100]: // TODO(minatoma): ignore blank lines and "# ..." comments

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 6]
```diff
@@ -73,42 +63,43 @@ static void ExtractKytheFacts(
...
-  for (const auto filename :
-       verible::make_range(args.begin() + 1, args.end())) {
-    std::string content;
-    if (!verible::file::GetContents(filename, &content).ok()) {
-      exit_status = 1;
-      continue;
-    }
+  std::string content;
+  if (!verible::file::GetContents(args[1], &content).ok()) {
+    LOG(INFO) << "Erro while reading file: " << args[1];
```
Comment (fangism) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:91]: Error

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 7]
```diff
@@ -73,42 +63,43 @@ static void ExtractKytheFacts(
...
 int main(int argc, char** argv) {
   const auto usage =
       absl::StrCat("usage: ", argv[0], " [options] <file> [<file>...]\n",
                    R"(
 verilog_kythe_extractor is a simple command-line utility
-to extract kythe indexing facts from the given file.
+to extract kythe indexing facts from the given file list.
 
-Expected Input: verilog file.
+Expected Input: verilog file list which contains the path of verilog files.
```
Comment (fangism) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:79]: Clarify that these files are top-level translation units, and should exclude `` `include``-d files.

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 8]
```diff
@@ -73,42 +63,43 @@ static void ExtractKytheFacts(
...
-  // All positional arguments are file names.  Exclude program name.
-  for (const auto filename :
-       verible::make_range(args.begin() + 1, args.end())) {
-    std::string content;
-    if (!verible::file::GetContents(filename, &content).ok()) {
-      exit_status = 1;
-      continue;
-    }
+  std::string content;
+  if (!verible::file::GetContents(args[1], &content).ok()) {
```
Comment (fangism) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:90]: Just FYI: if you don't have a file list, you can still use shell to create one on-the-fly, using `-` as a filename.  This can be useful for convenient testing.

```shell
v-v-k-extractor - <<EOF
file1.sv
file2.sv
EOF
```

File: verilog/tools/kythe/testdata/more_testdata/multi_file_test/file_list.txt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+multi-file-1.sv
+multi-file-2.sv
+multi-file-3.sv
+multi-file-4.sv
+
```
Comment (fangism) [inline verilog/tools/kythe/testdata/more_testdata/multi_file_test/file_list.txt:5]: Is this blank line handled safely?

**Review by fangism — CHANGES_REQUESTED** [review]
I'm only 10% through reviewing, will resume later today.

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/my_module2.sv (hunk 3 above)
Comment (ivan444) [reply verilog/tools/kythe/testdata/more_testdata/include_file_test/my_module2.sv:1]: I'm not sure. Can you check the other Kythe indexer projects? E.g., C++ or Go indexer should have a similar concept.

File: common/util/file_util.h [hunk 10]
```diff
@@ -44,6 +44,10 @@ struct Directory {
 // empty string.
 absl::string_view Basename(absl::string_view filename);
 
+// Returns the part of the path after the final "/".  If there is no
+// "/" in the path, the result is the same as the input.
+absl::string_view Direname(absl::string_view filename);
```
Comment (ivan444) [inline common/util/file_util.h:49]: You mean "the path before", not "after"?
How about "return the directory that contains the file". And then mention the edge cases (what if it ends with "/", what about ".", "", "./"

File: common/util/file_util_test.cc [hunk 11]
```diff
@@ -35,6 +35,14 @@ TEST(FileUtil, Basename) {
   EXPECT_EQ(file::Basename(""), "");
 }
 
+TEST(FileUtil, Direname) {
+  EXPECT_EQ(file::Direname("/foo/bar/baz"), "/foo/bar");
+  EXPECT_EQ(file::Direname("foo/bar/baz"), "foo/bar");
+  EXPECT_EQ(file::Direname("/foo/bar/"), "/foo/bar");
+  EXPECT_EQ(file::Basename("/"), "");
+  EXPECT_EQ(file::Basename(""), "");
```
Comment (ivan444) [inline common/util/file_util_test.cc:43]: Why Basename? :) 

Also, check for "./" and "."

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 12]
```diff
@@ -75,15 +77,47 @@ IndexingFactNode ExtractOneFile(absl::string_view content,
...
+IndexingFactNode ExtractFiles(std::vector<std::string> ordered_file_list,
+                              int& exist_status,
+                              absl::string_view file_list_dir) {
+  // Create a node to hold the dirname of the ordered file list and group all
+  // the files and acts as a ordered file list of these files.
+  IndexingFactNode file_list_facts_tree(IndexingNodeData(
+      {Anchor(file_list_dir, 0, 0)}, IndexingFactType::kFileList));
+  std::set<std::string> extracted_files;
+
+  for (const auto filename : ordered_file_list) {
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:107]: make it auto& or replace "const auto" with absl::string_view

File: verilog/tools/kythe/indexing_facts_tree_extractor.h [hunk 13]
```diff
@@ -166,12 +173,25 @@ class IndexingFactsTreeExtractor : public verible::TreeContextVisitor {
...
-// Given a verilog file returns the extracted indexing facts tree.
-IndexingFactNode ExtractOneFile(absl::string_view content,
-                                absl::string_view filename, int& exit_status,
-                                bool& parse_ok);
+// Given the ordered SystemVerilog files, Extracts and returns the
+// IndexingFactsTree from the given files.
+// The returned Root will have the files as children and the will be ordered as
+// the given order.
+IndexingFactNode ExtractFiles(std::vector<std::string> ordered_file_list,
+                              int& exist_status,
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.h:193]: s/exist_status/exit_status/

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 14]
```diff
@@ -75,15 +77,47 @@ IndexingFactNode ExtractOneFile(absl::string_view content,
...
+// Returns the path of the file relative to the file list.
+std::string GetPathRelativeToFileList(absl::string_view file_list_dir,
+                                      absl::string_view filename) {
+  return verible::file::JoinPath(file_list_dir, filename);
+}
+
+}  // namespace
+
+IndexingFactNode ExtractFiles(std::vector<std::string> ordered_file_list,
+                              int& exist_status,
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:99]: exit_status

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 15]
```diff
@@ -75,15 +77,47 @@ IndexingFactNode ExtractOneFile(absl::string_view content,
...
+
+// Returns the path of the file relative to the file list.
+std::string GetPathRelativeToFileList(absl::string_view file_list_dir,
+                                      absl::string_view filename) {
+  return verible::file::JoinPath(file_list_dir, filename);
+}
+
+}  // namespace
+
+IndexingFactNode ExtractFiles(std::vector<std::string> ordered_file_list,
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:98]: cost vector&

File: verilog/tools/kythe/kythe_facts_extractor.cc [hunk 16]
```diff
@@ -257,6 +289,34 @@ void KytheFactsExtractor::Visit(const IndexingFactNode& node) {
...
+    // vector so that it can outlive KytheFactsExtractor.
+    // The ScopeResolver-s are created and linked together as a linked-list
+    // structure so that the current ScopeResolver can search for definitions in
+    // the previous files' scopes.
+    std::string file_path = GetFilePathFromRoot(root);
+    if (scope_resolvers.empty()) {
+      scope_resolvers.push_back(ScopeResolver(nullptr));
+    } else {
+      // TODO(minatoma): fix this to pass a pointer to the last scope_resolver.
+      scope_resolvers.push_back(ScopeResolver(scope_resolvers.back()));
```
Comment (ivan444) [inline verilog/tools/kythe/kythe_facts_extractor.cc:306]: As we discussed in the morning, this is not doing what you want. It's a latent bug.

Can you look into fixing it in this PR?

File: verilog/tools/kythe/kythe_facts_extractor.cc [hunk 17]
```diff
@@ -717,34 +805,5 @@ std::ostream& operator<<(std::ostream& stream,
   return stream;
 }
 
-KytheIndexingData CollectKytheIndexingData(
```
Comment (ivan444) [inline verilog/tools/kythe/kythe_facts_extractor.cc:720]: @fangism we'll need to change the internal verification tests. It depends on this function.

@MinaToma can you provide a function similar to this one? (a function that returns complete KytheIndexingData)

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 18]
```diff
@@ -75,15 +77,47 @@ IndexingFactNode ExtractOneFile(absl::string_view content,
...
+  IndexingFactNode file_list_facts_tree(IndexingNodeData(
+      {Anchor(file_list_dir, 0, 0)}, IndexingFactType::kFileList));
+  std::set<std::string> extracted_files;
+
+  for (const auto filename : ordered_file_list) {
+    std::string content;
+    std::string file_path = GetPathRelativeToFileList(file_list_dir, filename);
+    if (!verible::file::GetContents(file_path, &content).ok()) {
+      LOG(INFO) << "Erro while reading file: " << file_path;
+      exist_status = 1;
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:112]: Might make sense to fully drop this status. The current code reports failure if any of the files error at reading. Is that intended?

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 19]
```diff
@@ -75,15 +77,47 @@ IndexingFactNode ExtractOneFile(absl::string_view content,
...
+  // the files and acts as a ordered file list of these files.
+  IndexingFactNode file_list_facts_tree(IndexingNodeData(
+      {Anchor(file_list_dir, 0, 0)}, IndexingFactType::kFileList));
+  std::set<std::string> extracted_files;
+
+  for (const auto filename : ordered_file_list) {
+    std::string content;
+    std::string file_path = GetPathRelativeToFileList(file_list_dir, filename);
+    if (!verible::file::GetContents(file_path, &content).ok()) {
+      LOG(INFO) << "Erro while reading file: " << file_path;
```
Comment (ivan444) [inline verilog/tools/kythe/indexing_facts_tree_extractor.cc:111]: s/INFO/ERROR/

Also s/Erro/Error/

**Review by ivan444 — CHANGES_REQUESTED** [review]

File: verilog/tools/kythe/kythe_facts_extractor.cc [hunk 20]
```diff
@@ -257,6 +289,34 @@ void KytheFactsExtractor::Visit(const IndexingFactNode& node) {
...
+  for (const IndexingFactNode& root : file_list.Children()) {
+    // Create a new ScopeResolver and give the ownership to the scope_resolvers
+    // vector so that it can outlive KytheFactsExtractor.
+    // The ScopeResolver-s are created and linked together as a linked-list
+    // structure so that the current ScopeResolver can search for definitions in
+    // the previous files' scopes.
+    std::string file_path = GetFilePathFromRoot(root);
+    if (scope_resolvers.empty()) {
+      scope_resolvers.push_back(ScopeResolver(nullptr));
+    } else {
```
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:304]: This is only true in the first iteration, could this be added outside of the for-loop?

File: verilog/tools/kythe/kythe_facts_extractor.cc [hunk 21]
```diff
@@ -649,6 +709,33 @@ VName KytheFactsExtractor::ExtractParamDeclaration(
...
+  const Scope* included_file_scope =
+      scope_resolver_->SearchForScope(Signature(file_path.Value()));
+  if (included_file_scope == nullptr) {
+    LOG(INFO) << "File Scope Not Found For file: " << file_path.Value();
+    return;
+  }
+
+  // Create child of edge between the parent and the member of the included
+  // file.
+  for (const ScopeMemberItem& item : included_file_scope->Members()) {
```
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:731]: s/item/member/

File: verilog/tools/kythe/kythe_facts_extractor.cc [hunk 22]
```diff
@@ -696,18 +783,19 @@ const std::set<Edge>& KytheFactsExtractor::GetExtractedEdges() const {
...
-std::string GetFilePathFromRoot(const IndexingFactNode& root) {
-  return root.Value().Anchors()[0].Value();
-}
-
 std::ostream& KytheFactsPrinter::Print(std::ostream& stream) const {
-  auto indexing_data = CollectKytheIndexingData(trees_);
-  for (const Fact& fact : indexing_data.facts) {
+  ScopeResolver scope_resolver(nullptr);
+  KytheFactsExtractor kythe_extractor(
+      GetFileListDirFromRoot(file_list_facts_tree_), &scope_resolver);
```
Comment (fangism) [inline verilog/tools/kythe/kythe_facts_extractor.cc:789]: I'm still a little concerned that printing requires passing a mutable object to the extractor, but perhaps I need to wrap my head around everything that is happening.  Can you add some comments about this? or point me to comments if you already have?

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/module.sv [hunk 23]
```diff
@@ -0,0 +1,28 @@
+`include "included-file-2.svh"
```
Comment (fangism) [inline verilog/tools/kythe/testdata/more_testdata/include_file_test/module.sv:1]: Should this come with some `//-` references to the included file?
And should the included files define themselves as some sort of anchor?

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/package_with_include.sv [hunk 24]
```diff
@@ -13,21 +13,15 @@ module my_module;
     //- @my_class1 ref MyClass1
     static my_class1 handle1 = new();
 
-    //- @my_class2 ref MyClass2
-    static my_class2 handle2 = new();
-
     //- @my_var ref MyVar
-    handle1.my_var = 10;
+    handle1::my_var = 10;
```
Comment (fangism) [inline verilog/tools/kythe/testdata/more_testdata/include_file_test/package_with_include.sv:17]: Just an observation: the way I think of chained hierarchical references is like this:

`A.B`: A is resolved with respect to current context, B is resolved w.r.t. **`typeof(A)`**'s type scope

`A::B`:  A is resolved with respect to current context, B is resolved w.r.t. **`A`**'s scope

**Review by fangism — CHANGES_REQUESTED** [review]
What happens if a file in the file list appears twice, duplicate?

File: verilog/tools/kythe/kythe_facts_extractor.cc (hunk 17 above)
Comment (fangism) [reply verilog/tools/kythe/kythe_facts_extractor.cc:720]: Ack.  @ivan444 I'll ask you to handle the PR import for this one.  I can share instructions offline.

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc (hunk 18 above)
Comment (fangism) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:112]: I suggested in a different comment to keep per-file information in an associative container/cache.  Yes, operation should still be robust enough to continue past file/IO errors, and just lose some facts/edges.  It's ok if the overall exit status is 0, as long as there is a channel to collect and report non-fatal errors.

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc [hunk 25]
```diff
@@ -938,5 +976,47 @@ void IndexingFactsTreeExtractor::ExtractForInitialization(
...
+  if (!verible::file::GetContents(file_path, &content).ok()) {
+    LOG(INFO) << "Erro while reading file: " << file_path;
+    return;
+  }
+
+  // Check if this included file was extracted before.
+  if (extracted_files_.find(file_path) != extracted_files_.end()) {
+    return;
+  }
+  extracted_files_.insert(file_path);
```
[the remark this answers is not in the document]
Comment (MinaToma) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:1015]: But do we need all this information here? 
We only check if we encountered this file before and every file is processed only once.

File: verilog/tools/kythe/testdata/more_testdata/multi_file_test/file_list.txt (hunk 9 above)
Comment (MinaToma) [reply verilog/tools/kythe/testdata/more_testdata/multi_file_test/file_list.txt:5]: stringstream ignores blank lines.

File: verilog/tools/kythe/indexing_facts_tree_extractor.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (MinaToma) [reply verilog/tools/kythe/indexing_facts_tree_extractor.cc:1015]: We extract each file from the file list in order and if we encountered an include we stop our extraction and extract the included file first, finish it, add it to the file list node and proceed with our file that included this file.

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/package_with_include.sv (hunk 24 above)
Comment (MinaToma) [reply verilog/tools/kythe/testdata/more_testdata/include_file_test/package_with_include.sv:17]: Can you give me an example I don't fully get what you are trying to say?

Comment (fangism) [reply verilog/tools/kythe/testdata/more_testdata/include_file_test/package_with_include.sv:17]: (no need to change anything)

This is just my thinking out loud about ways to implement member resolution using a CST traversal.

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/file_list.txt [hunk 26]
```diff
@@ -0,0 +1,3 @@
+module.sv
```
[the remark this answers is not in the document]
Comment (MinaToma) [reply verilog/tools/kythe/testdata/more_testdata/include_file_test/file_list.txt:1]: But how will file B see A?
`localparam B = A +1;`

**Comment by MinaToma** [comment]
> As we discussed in the morning, this is not doing what you want. It's a latent bug.
> 
> Can you look into fixing it in this PR?

still looking for a solution.

File: verilog/tools/kythe/testdata/more_testdata/include_file_test/file_list.txt (hunk 26 above)
[the remark this answers is not in the document]
Comment (MinaToma) [reply verilog/tools/kythe/testdata/more_testdata/include_file_test/file_list.txt:2]: 1- if the files are in wrong order no definitions will be found.
2- All the files (passed in the ordered file list or included) are processed only once.

File: verilog/tools/kythe/verilog_kythe_extractor.cc [hunk 27]
```diff
@@ -37,34 +37,24 @@ ABSL_FLAG(bool, printkythefacts, false,
...
-              << (!parse_ok ? " (incomplete due to syntax errors): " : "")
-              << std::endl;
-
-    std::cout << facts_tree << std::endl;
+    std::cout << file_list_facts_tree << std::endl;
   }
-  LOG(INFO) << '\n' << facts_tree;
-
-  indexing_facts_trees.push_back(facts_tree);
+  // LOG(INFO) << '\n' << file_list_facts_tree;
```
Comment (ivan444) [inline verilog/tools/kythe/verilog_kythe_extractor.cc:52]: Drop?

**Review by ivan444 — APPROVED** [review]

**Comment by MinaToma** [comment]
@fangism Ready to merge.

**Comment by MinaToma** [comment]
@fangism is this failed check is blocking?

**Review by fangism — APPROVED** [review]
Approving now, so we can give @ivan444 a chance to make the necessary fixes in PR import.

In follow-ups I'd like to see:

* re-organization into smaller demonstrative tests (unit-test size examples) into as many subdirectories as you like.
* possible refactoring/restructuring around multi-file handling.
